chess-AI/ai_io.py

```diff
@@ -2,63 +2,136 @@
-    """Save given model parameters to external file
+    The save_dir refers to the directory where the file
+    will be saved. For saving to Figshare, the file still needs to
+    be saved locally first.
     """
-    if save_path is not None:
-        torch.save(nnet.state_dict(), save_path)
-    else:
-        # Iterate through the number of models saved
-        for i in range(num_saved_models):
-            if not os.path.isfile(f'./models/models-{i + 1}.pt'):
-                if overwrite and i != 0:
-                    torch.save(nnet.state_dict(), f'./models/models-{i}.pt')
-                    break
-                torch.save(nnet.state_dict(), f'./models/models-{i + 1}.pt')
-                break
-            if i == num_saved_models - 1:
-                torch.save(nnet.state_dict(), f'./models/models-{num_saved_models}.pt')
-
-
-def load_model(nnet, model_path, num_saved_models):
-    """Load model parameters into given network from external file
+    # make sure a path was specified
+    date_string = create_date_string()
+    # get the full file path to save
+    full_path = ds_saving['data_dir'] + "dataset-" +  date_string + ".pt"
+    # save the dataset
+    torch.save(dataset, full_path)
+
+    if ds_saving['figshare_save']:
+        # Save dataset to figshare
+        title = "dataset-" + date_string
+        desc = "This description is a placeholder"
+        keys = ["Dataset", "Chess"]
+        # 179 is category [Artificial Intelligence and Image Processing]
+        categories = [179]
+        figshare_api.upload(title, desc, keys, categories, full_path)
+
+
+def save_model(nnet, m_saving, checkpointing, file_name=None):
+    """Save a model to Figshare or Locally.
+
+    The save_dir refers to the directory where the file will be saved. For saving to Figshare,
+    the file still needs to be saved locally first.
     """
-    if model_path is not None:
-        nnet.load_state_dict(torch.load(model_path))
+
+    if checkpointing:
+        assert file_name is not None, " For checkpointing, a file name msut be specified"
+        full_path = m_saving['checkpoint_path'] + file_name + ".pt"
+        torch.save(nnet.state_dict(), full_path)
     else:
-        for i in range(num_saved_models):
-            if not os.path.isfile(f'./models-{i + 2}.pt'):
-                if i != 0:
-                    nnet.load_state_dict(torch.load(f'./models-{i + 1}.pt'))
-                break
+        date_string = create_date_string()
+        full_path = m_saving['model_dir'] + "model-" +  date_string + ".pt"
+        torch.save(nnet.state_dict(), full_path)
+        # Check if model should be saved to Figshare
+        if m_saving['figshare_save']:
+            title = "model-" + date_string
+            desc = "This description is a placeholder"
+            keys = ["Neural Network", "Chess"]
+            # 179 is category, Artificial Intelligence and Image Processing
+            categories = [179]
+            api = FigshareApi()
```

Comment:
When you make `FigshareApi` a static class, please change this line to just use the class instead of instantiating an instance of it.